### PR TITLE
[REEF-634] Add an object that encapsulates information needed to recover an evaluator on restart

### PR DESCRIPTION
**Pull request by afchung**
This addressed the issue by
  * Creating ResourceRecoverEvent and changing EvaluatorRestartInfo to save the state of (potentially) restarted Evaluator.
  * Creating the interface ResourceEvent and have ResourceAllocationEvent and ResourceRecoverEvent extend it.
  * Have DriverRestartManager remember the status and other information of the evaluators to recover/fail on restart.

JIRA:
  [REEF-634](https://issues.apache.org/jira/browse/REEF-634)

### CODE REVIEW

**Comment by markusweimer** [comment]
I'm doing a pass.

**Comment by markusweimer** [comment]
`ResourceAllocationEventImpl` and `ResourceRecoverEventImpl` seem to be identical. Maybe we can share much code between them?

File: lang/java/reef-common/src/main/java/org/apache/reef/driver/restart/DriverRuntimeRestartManager.java [hunk 1]
```diff
@@ -58,9 +59,9 @@
 
   /**
    * Gets the sets of alive and failed evaluators based on the runtime implementation.
-   * @return EvaluatorRestartInfo, which encapsulates the alive and failed set of evaluator IDs.
+   * @return A map which encapsulates the states of previous evaluators.
    */
-  EvaluatorRestartInfo getAliveAndFailedEvaluators();
+  Map<String, EvaluatorRestartInfo> getPreviousEvaluators();
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/driver/restart/DriverRuntimeRestartManager.java]: Access to this Map happens only in `DriverRestartManager`, where it is used in the various `get()` methods. E.g. the `getFailedEvaluators()` call. Would it make sense to capture all that structure in its own collection object, e.g. `RestartEvaluators` which has access methods for the various conditions?

File: lang/java/reef-common/src/main/java/org/apache/reef/driver/restart/EvaluatorRestartInfo.java [hunk 2]
```diff
@@ -21,36 +21,68 @@
...
+  public boolean setEvaluatorRestartState(final EvaluatorRestartState to) {
+    if (EvaluatorRestartState.isLegalTransition(evaluatorRestartState, to)) {
+      this.evaluatorRestartState = to;
+      return true;
+    }
+
+    return false;
+  }
+
+  private EvaluatorRestartInfo(final Optional<ResourceRecoverEvent> resourceRecoverEvent,
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/driver/restart/EvaluatorRestartInfo.java]: Constructors should go to the top of the method list as per our coding style

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/YarnDriverRuntimeRestartManager.java [hunk 3]
```diff
@@ -64,11 +66,13 @@ private YarnDriverRuntimeRestartManager(@Parameter(YarnEvaluatorPreserver.class)
...
                                           final REEFEventHandlers reefEventHandlers,
                                           final ApplicationMasterRegistration registration,
-                                          final YarnContainerManager yarnContainerManager) {
+                                          final YarnContainerManager yarnContainerManager,
+                                          final RackNameFormatter rackNameFormatter) {
     this.registration = registration;
     this.evaluatorPreserver = evaluatorPreserver;
     this.reefEventHandlers = reefEventHandlers;
     this.yarnContainerManager = yarnContainerManager;
+    this.rackNameFormatter = rackNameFormatter;
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/YarnDriverRuntimeRestartManager.java]: Odd. Wasn't this added in an earlier PR?

**Comment by markusweimer** [comment]
@afchung I've done an initial pass.

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/YarnDriverRuntimeRestartManager.java (hunk 3 above)
Comment (afchung) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/YarnDriverRuntimeRestartManager.java]: Nope, that was in ``YarnContainerManager``. The need to format rackname only came in this PR.

**Comment by afchung** [comment]
@markusweimer Updated. Thanks!

**Comment by afchung** [comment]
@markusweimer Ran tests on YARN this time. They all passed. Thanks!

**Comment by markusweimer** [comment]
I'll do another pass.

**Pull request closed without merging**